providers/base/tests/test_wwan_tests.py

```diff
@@ -0,0 +1,203 @@
+        calls = [call(1), call(2)]
+
+        mock_get_equipment_id.assert_has_calls(calls)
+        mock_get_manufacturer.assert_has_calls(calls)
+        mock_get_model_name.assert_has_calls(calls)
+        mock_get_firmware_revision.assert_has_calls(calls)
+        mock_get_hardware_revision.assert_has_calls(calls)
+
+        resps = self._sorted_resouce_data(stdout.getvalue())
+        self.assertEqual(len(resps), 2)
+
+        for i in range(2):
+            self.assertEqual(set(resps[i].keys()), set(values.keys()))
+            for key in values.keys():
+                self.assertEqual(str(values[key][i]), resps[i][key])
+
+    @patch("wwan_tests.MMDbus.__init__")
+    @patch("wwan_tests.MMDbus.get_hardware_revision")
+    @patch("wwan_tests.MMDbus.get_firmware_revision")
+    @patch("wwan_tests.MMDbus.get_model_name")
+    @patch("wwan_tests.MMDbus.get_manufacturer")
+    @patch("wwan_tests.MMDbus.get_equipment_id")
+    @patch("wwan_tests.MMDbus.get_modem_ids")
+    def test_invoked_with_mmdbus(
+            self, mock_get_modem_ids, mock_get_equipment_id,
+            mock_get_manufacturer, mock_get_model_name,
+            mock_get_firmware_revision, mock_get_hardware_revision,
+            mock_mmdbus_init):
+
+        values = {
+            "mm_id": [1, 2],
+            "hw_id": ["d43f035387373e", "335464"],
+            "manufacturer": ["generic", "sierra"],
+            "model": ["MBIM [14C3:4D75]", "MBIM [14C3:9999]"],
+            "firmware_revision": [
+                "81600.0000.00.29.21.24_GC",
+                "81600.0000.00.29.21.25_GC"],
+            "hardware_revision": ["V1.0.6", "V1.0.7"]
+        }
+
+        mock_mmdbus_init.return_value = None
+        mock_get_modem_ids.return_value = values["mm_id"]
+        mock_get_equipment_id.side_effect = values["hw_id"]
+        mock_get_manufacturer.side_effect = values["manufacturer"]
+        mock_get_model_name.side_effect = values["model"]
+        mock_get_firmware_revision.side_effect = values["firmware_revision"]
+        mock_get_hardware_revision.side_effect = values["hardware_revision"]
+
+        sys.argv = ["wwan_tests.py", "resources"]
+        with redirect_stdout(StringIO()) as stdout:
+            wwan_tests.Resources().invoked()
+
+        mock_mmdbus_init.assert_called_with()
+        mock_get_modem_ids.assert_called_with()
+        calls = [call(1), call(2)]
+
+        mock_get_equipment_id.assert_has_calls(calls)
+        mock_get_manufacturer.assert_has_calls(calls)
+        mock_get_model_name.assert_has_calls(calls)
+        mock_get_firmware_revision.assert_has_calls(calls)
+        mock_get_hardware_revision.assert_has_calls(calls)
+
+        resps = self._sorted_resouce_data(stdout.getvalue())
+        self.assertEqual(len(resps), 2)
+
+        for i in range(2):
+            self.assertEqual(set(resps[i].keys()), set(values.keys()))
+            for key in values.keys():
+                self.assertEqual(str(values[key][i]), resps[i][key])
```

Comment:
This whole class is probably overkill.

Instead of having to mock so many things in `test_invoked_with_*`, you could go with:

```python
    @patch("wwan_tests.MMCLI")
    def test_invoked_with_mmcli(self, mock_mmcli):
        mmcli_instance = mock_mmcli.return_value
        mmcli_instance.get_modem_ids.return_value = ["test",]
        sys.argv = ["wwan_tests.py", "resources", "--use-cli"]
        wwan_tests.Resources().invoked()
        self.assertTrue(mock_mmcli.called)
        self.assertTrue(mmcli_instance.get_equipment_id.called)
        self.assertTrue(mmcli_instance.get_manufacturer.called)
        self.assertTrue(mmcli_instance.get_model_name.called)
        self.assertTrue(mmcli_instance.get_firmware_revision.called)
        self.assertTrue(mmcli_instance.get_hardware_revision.called)

    @patch("wwan_tests.MMDbus")
    def test_invoked_with_mmdbus(self, mock_mmdbus):
        mmdbus_instance = mock_mmdbus.return_value
        mmdbus_instance.get_modem_ids.return_value = ["test",]
        sys.argv = ["wwan_tests.py", "resources"]
        wwan_tests.Resources().invoked()
        wwan_tests.Resources().invoked()
        self.assertTrue(mock_mmdbus.called)
        self.assertTrue(mmdbus_instance.get_equipment_id.called)
        self.assertTrue(mmdbus_instance.get_manufacturer.called)
        self.assertTrue(mmdbus_instance.get_model_name.called)
        self.assertTrue(mmdbus_instance.get_firmware_revision.called)
        self.assertTrue(mmdbus_instance.get_hardware_revision.called)
```

I believe this would cover everything in your patch!

Reply:
This way, you don't need the `_sorted_resouce_data()` function either, by the way :)

Reply:
Thanks for the suggestions, I have modified the unit tests for the Resource class.